File: lab-3/tema-lab-2/t2-java/src/main/java/MultiThreaded.java
import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Scanner;
import java.util.concurrent.BrokenBarrierException;
import java.util.concurrent.CyclicBarrier;

public class MultiThreaded {
    public static void main(String[] args) {
        try {
            // reading
            File file = new File(args[0]);
            Scanner reader = new Scanner(file);
            int p = Integer.parseInt(args[1]);
            String[] words;

            int N = Integer.parseInt(reader.nextLine());
            int M = Integer.parseInt(reader.nextLine());
            int[][] F = new int[N][M];
            for (int i = 0; i < N; i++) {
                words = reader.nextLine().split(" ");
                for (int j = 0; j < M; j++) {
                    F[i][j] = Integer.parseInt(words[j]);
                }
            }

            int n = Integer.parseInt(reader.nextLine());
            int m = Integer.parseInt(reader.nextLine());
            int[][] W = new int[n][m];
            for (int i = 0; i < n; i++) {
                words = reader.nextLine().split(" ");
                for (int j = 0; j < m; j++) {
                    W[i][j] = Integer.parseInt(words[j]);
                }
            }
            reader.close();

            // multithreading init
            Thread[] threads = new Thread[p];
            CyclicBarrier barrier = new CyclicBarrier(p);
            int startI = 0, stopI, batchSize = N / p, surplus = N % p;

            // applying kernel
            Timer.Start();
            for (int i = 0; i < p; i++) {
                stopI = startI + batchSize + (surplus > 0 ? 1 : 0);
                threads[i] = new Thread(F, W, startI, stopI, N, M, n, m, barrier);
                threads[i].start();
                startI = stopI;
                surplus--;
            }

            // multithreading joining
            for (int i = 0; i < p; i++) {
                try {
                    threads[i].join();
                } catch (InterruptedException ex) {
                    ex.printStackTrace();
                }
            }
            System.out.println(Timer.Stop());
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private static class Thread extends java.lang.Thread {
        private final int[][] F;
        private final int[][] W;
        private final int startI, stopI, N, M, n, m;
        private final CyclicBarrier barrier;

        public Thread(int[][] F, int[][] W, int startI, int stopI, int N, int M, int n, int m, CyclicBarrier barrier) {
            this.F = F;
            this.W = W;
            this.startI = startI;
            this.stopI = stopI;
            this.N = N;
            this.M = M;
            this.n = n;
            this.m = m;
            this.barrier = barrier;
        }

        @Override
        public void run() {
            try {
                // copying the needed values
                int originN = startI - (n / 2), originM = -(m / 2), sizeI = (stopI - startI) + (n - 1), sizeJ = M + (m - 1);
                int[][] aux = new int[sizeI][sizeJ];
                for (int i = originN; i < originN + sizeI; i++) {
                    for (int j = originM; j < originM + sizeJ; j++) {
                        int difI = i;
                        if (difI <= -1) difI = 0;
                        else if (difI >= N) difI = N - 1;
                        int difJ = j;
                        if (difJ <= -1) difJ = 0;
                        else if (difJ >= M) difJ = M - 1;
                        aux[i + (n / 2) - startI][j + (m / 2)] = F[difI][difJ];
                    }
                }

                // waiting for all the other threads
                barrier.await();

                // applying the filter
                int wI, wJ;
                for (int i = startI; i < stopI; i++) {
                    for (int j = 0; j < M; j++) {
                        int newValue = 0;
                        wI = 0;
                        for (int k = i - startI; k < i - startI + n; k++) {
                            wJ = 0;
                            for (int l = j; l < j + m; l++) {
                                newValue += aux[k][l] * W[wI][wJ++];
                            }
                            wI++;
                        }
                        F[i][j] = newValue;
                    }
                }
            } catch (InterruptedException | BrokenBarrierException e) {
                e.printStackTrace();
            }
        }
    }
}

File: lab-3/tema-lab-2/t2-cpp/src/multiThreaded.cpp
#include "multiThreaded.h"
#include "timer.h"

using namespace std;

struct threadParameters {
    int** F;
    int** W;
    int startI;
    int stopI;
    int N;
    int M;
    int n;
    int m;
    barrier* myBarrier;
};

void run(threadParameters params) {
    int originN = params.startI - (params.n / 2), originM = -(params.m / 2), sizeI = (params.stopI - params.startI) + (params.n - 1), sizeJ = params.M + (params.m - 1);
    int** aux = new int*[sizeI];
    for (int i = 0; i < sizeI; ++i) {
        aux[i] = new int[sizeJ];
    }
    for (int i = originN; i < originN + sizeI; i++) {
        for (int j = originM; j < originM + sizeJ; j++) {
            int difI = i;
            if (difI <= -1) difI = 0;
            else if (difI >= params.N) difI = params.N - 1;
            int difJ = j;
            if (difJ <= -1) difJ = 0;
            else if (difJ >= params.M) difJ = params.M - 1;
            aux[i + (params.n / 2) - params.startI][j + (params.m / 2)] = params.F[difI][difJ];
        }
    }

    // waiting for all the other threads
    params.myBarrier->wait();

    // applying the filter
    int wI, wJ;
    for (int i = params.startI; i < params.stopI; i++) {
        for (int j = 0; j < params.M; j++) {
            int newValue = 0;
            wI = 0;
            for (int k = i - params.startI; k < i - params.startI + params.n; k++) {
                wJ = 0;
                for (int l = j; l < j + params.m; l++) {
                    newValue += aux[k][l] * params.W[wI][wJ++];
                }
                wI++;
            }
            params.F[i][j] = newValue;
        }
    }

    // memory cleanup
    for (int i = 0; i < sizeI; ++i) {
        delete[] aux[i];
    }
    delete[] aux;
}

void multiThreaded(int argc, char **argv) {
    ifstream fin(argv[1]);
    string line;
    string word;
    Timer timer = Timer();
    int p = stoi(argv[2]), N, M, n, m;

    // reading
    getline(fin, line);
    N = stoi(line);
    getline(fin, line);
    M = stoi(line);
    int** F = new int*[N];
    for (int i = 0; i < N; ++i) {
        F[i] = new int[M];
        for (int j = 0; j < M; j++) {
            getline(fin, word, ' ');
            F[i][j] = stoi(word);
        }
    }

    getline(fin, line);
    getline(fin, line);
    n = stoi(line);
    getline(fin, line);
    m = stoi(line);
    int** W = new int*[n];
    for (int i = 0; i < n; ++i) {
        W[i] = new int[m];
        for (int j = 0; j < m; ++j) {
            getline(fin, word, ' ');
            W[i][j] = stoi(word);
        }
    }

    // multithreading init
    thread threads[p];
    int startI = 0, stopI, batchSize = N / p, surplus = N % p;
    threadParameters params{};
    params.F = F;
    params.W = W;
    params.N = N;
    params.M = M;
    params.n = n;
    params.m = m;
    params.myBarrier = new barrier(p);

    // apply kernel
    timer.start();
    for (int i = 0; i < p; ++i) {
        stopI = startI + batchSize + (surplus > 0 ? 1 : 0);
        params.startI = startI;
        params.stopI = stopI;
        threads[i] = thread(run, params);
        startI = stopI;
        surplus--;
    }

    // multithreading joining
    for (int i = 0; i < p; i++) {
        threads[i].join();
    }

    // memory cleanup
    for (int i = 0; i < N; ++i) {
        delete[] F[i];
    }
    delete[] F;
    for (int i = 0; i < n; ++i) {
        delete[] W[i];
    }
    delete[] W;

    cout << timer.stop();
}